File: 03_LearnDSP-Python/Sec04_PracticalMethods/case02_EnvelopeExtraction.py
'''
Extract envelope curve of a modulated signal
Use Hilbert transform

XiaoCY 
'''

#%%
import numpy as np
import matplotlib.pyplot as plt
import scipy.signal as sig

fs = 100.
fsig = 1.

mu = 4.
b = mu/3.
t = np.arange(0,2*mu,1/fs)
evlp = np.exp(-(t-mu)**2/(2*b**2))
x = np.sin(2*np.pi*fsig*t)*evlp

#%% Hilbert Transform
y = sig.hilbert(x)

plt.figure()
ax = plt.axes(projection='3d')
ax.plot(t,x,label='signal')
ax.plot3D(t,y.real,y.imag,label='Hilbert')
ax.grid()
ax.legend()
ax.set_xlabel('Time (s)')
ax.set_ylabel('Re(H)')
ax.set_zlabel('Im(H)')
plt.show()

plt.figure
plt.plot(t,x,label='signal')
plt.plot(t,np.abs(y),label='Hilbert')
plt.plot(t,evlp,'--',label='envelope')
plt.grid()
plt.legend()
plt.xlabel('Time (s)')
plt.ylabel('Signal')
plt.show()

#%% Transform with bias
bias = 1.
x1 = x+bias
evlp1 = evlp+bias

y1 = sig.hilbert(x1)
xm = np.mean(x1)
y2 = sig.hilbert(x1-xm)

plt.figure()
plt.plot(t,x1,label='signal')
plt.plot(t,np.abs(y1),label='direct Hilbert')
plt.plot(t,np.abs(y2)+xm,label='bias-cancelled ilbert')
plt.plot(t,evlp1,'--',label='envelope')
plt.grid()
plt.legend()
plt.xlabel('Time (s)')
plt.ylabel('Signal')
plt.show()

File: 02_LearnDSP-MATLAB/Sec04_PracticalMethods/case02_EnvelopeExtraction.m
% Extract envelope curve of a modulated signal
% Use Hilbert transform

% XiaoCY 2021-02-14

%%
clear;clc

fs = 100;
fsig = 1;

mu = 4;
b = mu/3;
t = 0:1/fs:2*mu;
evlp = exp(-(t-mu).^2/(2*b^2));
x = sin(2*pi*fsig*t).*evlp;

%% Hilbert transform
y = hilbert(x);

figure
plot(t,x,'DisplayName','signal')
hold on
grid on
plot3(t,real(y),imag(y),'DisplayName','Hilbert')
legend
xlabel('Time (s)')
ylabel('Re(H)')
zlabel('Im(H)')
view(10,45)

figure
plot(t,x,'DisplayName','signal')
hold on
grid on
legend
plot(t,abs(y),'DisplayName','Hilbert')
plot(t,evlp,'--','DisplayName','envelope')
xlabel('Time (s)')
ylabel('Signal')

%% Transform with bias
bias = 1;
x1 = x+bias;
evlp1 = evlp+bias;

y1 = hilbert(x1);
xm = mean(x1);
y2 = hilbert(x1-xm);

figure
plot(t,x1,'DisplayName','signal')
hold on
grid on
legend
plot(t,abs(y1),'DisplayName','direct Hilbert')
plot(t,abs(y2)+xm,'DisplayName','bias-cancelled Hilbert')
plot(t,evlp1,'--','DisplayName','envelope')
xlabel('Time (s)')
ylabel('Signal')